aptos-move/move-examples/common_account/sources/common_account.move

```diff
@@ -0,0 +1,269 @@
+    use aptos_framework::account::{Self, SignerCapability};
+
+    /// Missing the common account Management
+    const ENO_MANAGEMENT_RESOURCE_FOUND: u64 = 1;
+    /// Missing CommonAccount
+    const ENO_ACCOUNT_RESOURCE_FOUND: u64 = 2;
+    /// Missing the common account Capability
+    const ENO_CAPABILITY_FOUND: u64 = 3;
+    /// Was not offered a capability
+    const ENO_CAPABILITY_OFFERED: u64 = 4;
+    /// Signer isn't admin
+    const ENOT_ADMIN: u64 = 5;
+    /// Found an address different than expected at the Capability
+    const EUNEXPECTED_PARALLEL_ACCOUNT: u64 = 6;
+
+    /// Contains the signer capability that generates the common account signer.
+    struct CommonAccount has key {
+        signer_cap: SignerCapability,
+    }
+
+    struct Empty has drop, store {}
+
+    /// Contains the framework for managing the account, partiuclarly around adminstration.
+    struct Management has key {
+        admin: address,
+        unclaimed_capabilities: SimpleMap<address, Empty>,
+    }
+
+    /// A revokable capability that is stored on a users account.
+    struct Capability has drop, key {
+        common_account: address,
+    }
+
+    /// Creates a new common account by creating a resource account and storing the capability.
+    public entry fun create(sender: &signer, seed: vector<u8>) {
+        let (resource_signer, signer_cap) = account::create_resource_account(sender, seed);
+
+        move_to(
+            &resource_signer,
+            Management {
+                admin: signer::address_of(sender),
+                unclaimed_capabilities: simple_map::create(),
+            },
+        );
+
+        move_to(&resource_signer, CommonAccount { signer_cap });
+    }
+
+    /// Add the other account to the list of accounts eligible to claim a capability for this
+    /// common_account
+    public entry fun add_account(
+        sender: &signer,
+        common_account: address,
+        other: address,
+    ) acquires Management {
+        let management = assert_is_admin(sender, common_account);
+        simple_map::add(&mut management.unclaimed_capabilities, other, Empty {});
+    }
+
+    /// Remove the other account to the list of accounts eligible to claim a capability for this
+    /// common_account if it exists or remove the capability.
+    public entry fun remove_account(
+        sender: &signer,
+        common_account: address,
+        other: address,
+    ) acquires Capability, Management {
+        let management = assert_is_admin(sender, common_account);
+        if (simple_map::contains_key(&management.unclaimed_capabilities, &other)) {
+            simple_map::remove(&mut management.unclaimed_capabilities, &other);
+        } else {
+            assert!(exists<Capability>(other), error::not_found(ENO_CAPABILITY_FOUND));
+            move_from<Capability>(other);
+        }
+    }
+
+    /// Acquire the capability to use the signer capability for the common_account.
+    public entry fun acquire_capability(
```

Comment:
The workflow seems to be `create` -> `add_account` -> `acquire_capability` -> `acquire_signer`. I am wondering if we could skip the `acquire_capability` step.

I think that the practical reason for `acquire_capability` is to let the whitelisted user to send a transaction to publish a `Capability` resource under their account. Then, later, the permission check for the user can be done efficiently by checking `exists<Capability>(...)`. I am wondering if this is the only reason for `acquire_capability`?

Otherwise, `unclaimed_capabilities` could serve as an access control list. I am wondering if what is the data structure (among `smart_vector`, `smart_table`, `simple_map`) to efficiently implement this kind of access control list if any?

Reply:
Primarily to eliminate "expensive" ACL lookups every time a user wants to acquire the signer.

We could definitely move toward a `smart_table` instead of this but that has a layer of indirection immediately. `smart_vector` would embrace a potentially high cost.

Each solution has tradeoffs.

Another model would be to provide an off-chain cryptographic proof that let's the sender claim the capability.

Reply:
I think to your point, @junkil-park , we could actually tweak the work flow, where we merge `acquire_capability` into acquire_signer`